man/man1/make_coupler.1
.TH make_coupler 1 "atlc-4.0.0 13/4/02" "Dr. David Kirkby"
.ds n 5
.SH NAME
\fBmake_coupler\fR - bitmap generator for coupler with thin striplines between two infinitly wide groundplanes (part of \fBatlc\fR)
.SH SYNOPSIS
\fBmake_coupler [-C] [-b bitmapsize] [-v] H w s\fR
.br
.SH WARNING
\fBmake_coupler \fR prints binary output to the standard output and will
screw your terminal unless you redirect it. See the examples below. 
.SH DESCRIPTION
\fBmake_coupler\fR is a pre-processor for \fBatlc\fR, the finite difference 
programme that is used to calculate the 
properties of a two and three conductor electrical transmission lines of arbitrary 
cross section. The programme \fBmake_coupler\fR is used as a fast way of
generating bitmaps (there is no need to use a graphics programme), for a
pair of 1 pixel thick conductors of width w, spaced a distance s apart, 
inside a pair of
groundplanes of  separation H. This has an exact analytical
answer, when the inner is infinitely thin and the overall width W is infinitely
wide. This theoretical impedance is calculated and printed to stderr if:
.br
a) The -v option is given on make_coupler's command line. AND
.br
b) make_coupler was linked against the gsl library at http://sources.redhat.com/gsl/
.P
-----------------------------------------------------  ^
.br
|                                                   |  |
.br                                                     
|                  Er=1.0                           |  |
.br                                                     
|                                                   |  |
.br                                                     
|            -----------       -----------          |  H
.br
|            <----w----><--s--><----w---->          |  |
.br       
|                                                   |  |
.br
|                                                   |  |
.br
|                                                   |  |
.br
-----------------------------------------------------  v
.br
<-------------------------W------------------------->
.br

The overall width W is calculated internally to be large enough that
making it any larger will have negligable effect on the results.
The dielectric is assumed that of vacuum.
.PP 
.br
The bitmaps produced by \fBmake_coupler\fR are 24-bit bit colour bitmaps, as are required by \fBatlc\fR. Currenly make_coupler only produces bitmaps with
white (vacuum) dielectrics. 

.SH OPTIONS
.PP
\fB-C \fR
.br
Print the full copyright information and exit.
.PP
.br
\fB-b bitmapsize\fR
.br
is used to set the size of the bitmap, and so the accuracy to which atlc
is able to calculate the coupler's properties. The default
value for 'bitmapsize' is normally 18 although this can be set at
compile time. (see 'DEFAULT_COUPLER_BITMAP_SIZE' in definitions.h). The value can be set anywhere from 6 to 28. The size of bitmap produced will be
between 2^(bitmapsize-1) and 2^(bitmapsize+1), so if its set to 28 the
bitmap could be as big as 512 Mb, which would take an eternity to
compute. 
.PP
\fB-v \fR
.br
Print some extra data to stderr on the theoetical impedance of the
structure. This is useful to compare with the numerical result from
\fBatlc\fR
.SH EXAMPLES
.br
In the first example, a standard coupler is created, where the 


eg1. make_coupler  1.23 0.7 1.2 1. >   coupler.bmp
e.g. make_coupler -b 8 1.23 0.7 1.2 >  small.bmp   (small bitmap produced)
e.g. make_coupler -v 12 1.23 0.7 1.2 >  theory.bmp  (prints info to stderr)













.P
.SH SEE ALSO
atlc(1), rect_cen_in_rect(1), circ_in_circ(1). rect_in_rect(1), circ_in_rect(1).
rect_in_circ(1), readbin(1) and sym_strip(1).
.P 
.br
http://atlc.sourceforge.net                - Home page 
.br
http://sourceforge.net/projects/atlc       - Download area
.br
atlc-X.Y.Z/docs/html-docs/index.html       - HTML docs
.br
atlc-X.Y.Z/docs/qex-december-1996/atlc.pdf - theory paper
.br
atlc-X.Y.Z/examples                        - examples
.br
http://www.david-kirkby.co.uk              - my home page
.br
http://www.david-kirkby.co.uk/ham          - ham radio pages
.br
.SH AUTHOR
Dr. David Kirkby PhD, email drkirkby@ntlworld.com, G8WRB.
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (c) 2002 Dr. David Kirkby. 
.PP

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.
.pp
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.p
You should have received a copy of the GNU General Public License (in
the file COPYING) along with this program. You may get a full copy of it
by running the programme with the -C option. Otherwise you can write to the 
Free Software Foundation, Inc., 59 Temple Place - Suite 330, Boston, 
MA  02111-1307, USA.


NEWS
NEWS for realease 4.3.0 on 23 May 2003.

A bug, preventing atlc from runningt properly on multiple
processors has been fixed. This is the main change. 

atlc should no longer fail any test on a multi-processor
system configured with --with-threads. The algorithm used
for both single processor and multiple processor (i.e. 
threaded code is the same). 

The -t option to atlc, which sets the number of threads 
when configured for multiple processors can be set to 0
to use entirely the single-threaded algorithm. If set to
one, it will use the multi-threaded algoritm, but use
only one thread. If set to some other number, it will use
that number of threads and be optimal for the same number
of cpus. 

Support for distributed programming has been temporarily removed.
Someone else added the code, but since he is no longer involved in
the development, it is difficult for me to follow. So I'll remove
the MPI specific code and will re-implement this. 

A test has been added to make sure atlc can handle the -d option 


Dr. David Kirkby 23/5/2003



src/non_gui/make_coupler.c
/* atlc - arbitrary transmission line calculator, for the analysis of
transmission lines are directional couplers. 

Copyright (C) 2002. Dr. David Kirkby, PhD (G8WRB).

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either package_version 2
of the License, or (at your option) any later package_version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
USA.

Dr. David Kirkby, e-mail drkirkby@ntlworld.com 

*/

/* The following calculates the odd and even mode impedances between two
zero thickness strips of width w, spaced a distance s between two
groundplanes of spacing h. */

#define RATIO 8
#include "definitions.h"

#ifdef HAVE_STDLIB_H
#include <stdlib.h>
#endif

#ifdef HAVE_STRING_H
#include <string.h>
#endif

#ifdef HAVE_GSL_GSL_SF_ELLINT_H
#include <gsl/gsl_sf_ellint.h>
#endif

double WW, HH, ss, ww, Er1, Er2;
int W, H, s,  w,  verbose=FALSE;

int main(int argc, char **argv)
{
  double er, return_error;
  double Zodd, Zeven, Zo;
  int bmp_size=DEFAULT_COUPLER_BITMAP_SIZE, q; 
  FILE *image_data_fp;
  struct data optimise;

  while((q=get_options(argc,argv,"Cb:f:v")) != -1)
  switch (q) 
  {
    case 'C':
      print_copyright((char *) "2002");
      exit(1);
    break;
    case 'b':
      bmp_size=atol(my_optarg); 
    break;
    case 'v':
      verbose=TRUE;
    break;
  }
  if( bmp_size < 6 || (argc-my_optind != 4) || bmp_size>28)
  {
    usage_make_coupler();
    exit(1);
  }

  HH=atof(argv[my_optind]);
  ww=atof(argv[my_optind+1]);
  ss=atof(argv[my_optind+2]);
  //er=atof(argv[my_optind+3]);
  er=1.0;
  image_data_fp=fopen(argv[my_optind+3],"wb");
  if(er <1.0)
  {
    fprintf(stderr,"Er must be 1.0 or more \n");
    exit(1);
  }

  WW=2.0*ww+ss+RATIO*HH;
  optimise.float_values[0]=WW            ;   /* minimum width as a float*/
  optimise.float_values[1]=HH;               /* height in floats */
  optimise.float_values[2]=ww;               /* stripline width */
  optimise.float_values[3]=ss;               /* spacing between the strips */

  optimise.importance[0]=NOT_IMPORTANT;      /* W is non critical */
  optimise.importance[1]=MOST_IMPORTANT;          /* H is critical */
  optimise.importance[2]=IMPORTANT;          /* w is critical */
  optimise.importance[3]=IMPORTANT;     /* s is most critical */

  optimise.odd_or_even[0]=DONT_CARE;        /* W can be odd or even */
  optimise.odd_or_even[1]=ODD;              /* H must be even */
  optimise.odd_or_even[2]=DONT_CARE;        /* w can be odd or even */
  optimise.odd_or_even[3]=DONT_CARE;        /* s can be odd or even */

  /* We will now optimise for the 4 parameters W, H, w and s, forcing H
  to be odd, and not bothering what the width is, as long as its about
  right. */

  return_error=calculate_integer_values(&optimise, 4, bmp_size);
  if(verbose)
    printf("error returned from calculate_integer_values=%g\n", return_error);
  W=optimise.best[0];
  H=optimise.best[1];
  w=optimise.best[2];
  s=optimise.best[3];

  write_bitmap(image_data_fp);
  if(verbose)
  {
    calculate_Zodd_and_Zeven(&Zodd, &Zeven, &Zo, w, H-10, s, 1.0);
    printf("Wanted:            Zodd=%f Zeven=%f Zo=%f\n", Zodd, Zeven, Zo);

    calculate_Zodd_and_Zeven(&Zodd, &Zeven, &Zo, ww, HH, ss, 1.0);
    printf("Actually modelled: Zodd=%f Zeven=%f Zo=%f\n", Zodd, Zeven, Zo);
  }
  exit(0);
}
